Cleanup code and change to a more pythonic API

mgautierfr opened the pull request:
Big code clean, fix bugs and change API to a more pythonic one (and less prone to error).
This rewrite almost all the pyx file :)

This cause a API break, so it should be preferable to release a new major version.

mgautierfr committed:
Remove __setitem__ method from TagEntry.

This method is buggy and I don't see any use case where a user would like
to change those values from python code.

This method is buggy cause :
- Assigning string python object to 'char *' C variable do not copy but
  directly use the inner data array of the string. Once the string python
  object is garbage collected, the 'char *' points to garbage data.
- Assigning to the fields items is buggy cause the for loop set only the
  first item of the list (and after having set it to NULL)

This method has never work and so probably no one use it.
There will be no problem if we remove it.

mgautierfr committed:
Remove __cinit__ method from TagEntry.

The fields struct is already initialized by C code before parsing a tag
line (in the tags file) and setting the tagEntry. We do not need to do
it ourselves.

mgautierfr committed:
Respect 'const' qualifier.

Most 'char *' attributes, function arguments or return values are const.
Keep it in Cython.

In __getitem__, result should be 'const char*' now, but Cython is smart
enough to infer it itself. So no need to be explicit.

mgautierfr committed:
Raise KeyError if the key is not present in the tag.

Behave more link a dict, raise a KeyError instead of returning None.

mgautierfr committed:
We need to encode the string to pass it as const char* to ctagsField.

mgautierfr committed:
Raise KeyError if item is not in the Ctags file.

mgautierfr committed:
Return a new entry instead of taking an entry as input to change it.

Taking a argument as ouput is a C idiom. In python, we return the value.

The entry is a plain dict, not anymore a class with Mapping API.

More than the pythonic API, this fix also a bug :
The C tagEntry share the same internal buffer to store the datas.
ctags* function "just" changes the entries' pointers to the right place
in the buffer. Using one of those functions will invalidate any other
entries already returned. By using a dict we make a plain copy of the
string and so entry are now independent of the internal state of the C
code.

mgautierfr committed:
Use generators to explore the CTags file.

We do not use first/next/find* methods anymore.

'*next' methods are depends of the context set by other methods.
Hiding those methods in a high level API ensure that they are not badly used.

'first' and 'find' can respectively replace by next(ctagsfile.all_tags()) and
next(ctagsfile.find_tags(...))

mgautierfr committed:
Fix tests to use new api

mgautierfr committed:
Make TagEntry behave like a Mapping.

Use collections.abc.Mapping to let TagEntry be usable has a Mapping.

cdef class cannot inherit from a Python class. So rename it to cTagEntry
and use classic class TagEntry to do the Mixin.

# Conflicts:
#	README.rst

mgautierfr committed:
Raise Exception instead of returning status.

The common Python idiom is to raise Exception instead of returning status
values.

Move code of open method in __cinit__.

No more access to CTags attributes 'opened' and 'error_number'.
This is no more needed with the raise of OSError.

# Conflicts:
#	README.rst

sbraz commented on src/_readtags.pyx:
```diff
@@ -17,130 +17,142 @@ You should have received a copy of the GNU General Public License
-        if not self.info.status.opened:
-            raise Exception('Invalid tag file')
+        else:
+            if key == 'author':
```
I'd do something like this to further clean up the code

``` py
if key in ('author', 'name', 'url', 'version'):
    ret = getattr(self.info.program, key)
```

mgautierfr replied:
I'm not sure getattr works for c structure. Maybe cython is smart enough.
I will test.

mgautierfr replied:
Well, Cython is not smart enough :) getattr doesn't work on c structure.

sbraz commented on src/_readtags.pyx:
```diff
@@ -17,130 +17,142 @@ You should have received a copy of the GNU General Public License
+
+            WARNING: Only one iterator can run on a tag file.
+            If you use another iterator (by calling all_tags or find_tags),
+            any previous iterator will be invalidate and raise a RuntimeError.
```
→ invalidated

sbraz commented on README.md:
```diff
@@ -35,75 +35,104 @@ ctags --fields=afmikKlnsStz readtags.c  readtags.h
     tagFile = CTags('tags')
-except:
+except OSError as err:
+    print err
```
It would be nice to use Python-3-compatible syntax here.

mgautierfr replied:
The "except OSError as err" is python3 syntax.
Do you speak about the print function ?

sbraz replied:
Yep

On 17 May 2016 at 17:42, Matthieu Gautier notifications@github.com wrote:

> In README.md
> https://github.com/jonashaag/python-ctags3/pull/6#discussion_r63547497:
> 
> >  import sys
> > 
> >  try:
> >      tagFile = CTags('tags')
> > -except:
> > +except OSError as err:
> > -    print err
> 
> The "except OSError as err" is python3 syntax.
> Do you speak about the print function ?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> https://github.com/jonashaag/python-ctags3/pull/6/files/540b4bfd9dda55f4b823652dc507d5cbe9c60621#r63547497

sbraz commented:
That looks like a nice improvement.
Would it be possible to make the whole library take `str` as input instead of `bytes` and also have it return `str`?

mgautierfr commented:
I could be done for the path of the ctags file (using filesystem encoding).

However, for the tags entries it seems a bit more complicated : Which encoding use ?
The content of the tag file is probably dependent of the source files encoding and we cannot know which one it is. I think it is better to let the library return bytes and let the user code try to decode the bytes using the right encoding.

mgautierfr committed:
first and find method raise RuntimeError, not KeyError.

mgautierfr committed:
Accept bytes and str as tag file path.

mgautierfr committed:
Add few tests.

mgautierfr committed:
Update README (switch to python3 syntax for print)

jonashaag commented:
How about this: add an encoding parameter when loading the tag file that defaults to UTF8 and show people how to specify an encoding when generating the  ctags file in the readme.

mgautierfr committed:
Add a encoding arg to the CTags file to automatically decode file content.

Encoding argument to CTags file is used to automatically encode/decode
content of the CTags file.

mgautierfr commented:
Last commit add the encoding/encoding_errors parameter at CTags creation.
By default, utf8 is used and the ctags' entry contains string instead of bytes.

mgautierfr committed:
Use str instead of unicode to be compatible with python2.

jonashaag commented:
Could you please move the unrelated ckeanups like in the readme to another pull request? Otherwise it's just too much to review at once. Thanks

jonashaag commented:
What's the state of this?

sbraz commented:
@mgautierfr Hello, same question as Jonas, what's the state of this?

masatake commented:
> This cause a API break, so it should be preferable to release a new major version.

I think it is better to give a new name like python3-readtags, python-readtags, python3-libreadtags, or python3-libreadtags to the project.

masatake commented:
@mgautierfr can we take over this pull request?

jonashaag commented:
Yes, let's do it :)

mgautierfr commented:
Sorry my no response to first ping.

Yes, take over this PR. I don't use `python-ctags3` now so I will not work anymore on this.
You're free to continue my work (just keep some credits to me :) )
